File: bindings/aster_python/aster/contract/manifest.py
"""
aster.contract.manifest — ContractManifest and verification.

Spec reference: Aster-ContractIdentity.md §11.4

Provides:
- ContractManifest: dataclass for persisting contract identity info
- FatalContractMismatch: raised when a live contract doesn't match the manifest
- verify_manifest_or_fatal: strict identity check at startup
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field

import blake3


@dataclass
class ContractManifest:
    """Persisted record of a service contract's canonical identity.

    Write with ``json.dumps(asdict(manifest))`` and read back with
    ``ContractManifest(**json.loads(text))``.
    """

    service: str
    """Service name."""

    version: int
    """Service version integer."""

    contract_id: str
    """64-char hex string (full BLAKE3 digest of canonical ServiceContract bytes)."""

    canonical_encoding: str = "fory-xlang/0.15"
    """Encoding scheme identifier."""

    type_count: int = 0
    """Number of user-defined types referenced by this contract."""

    type_hashes: list[str] = field(default_factory=list)
    """Hex-encoded BLAKE3 hashes of each referenced TypeDef (sorted)."""

    method_count: int = 0
    """Number of methods in the service."""

    serialization_modes: list[str] = field(default_factory=list)
    """Supported serialization mode strings (e.g. ["xlang"])."""

    alpn: str = "aster/1"
    """ALPN protocol identifier."""

    scoped: str = "shared"
    """Service scope: "shared" or "stream"."""

    deprecated: bool = False
    """Whether this contract version is deprecated."""

    semver: str | None = None
    """Optional semantic version string."""

    vcs_revision: str | None = None
    """Optional VCS commit hash."""

    vcs_tag: str | None = None
    """Optional VCS tag."""

    vcs_url: str | None = None
    """Optional VCS repository URL."""

    changelog: str | None = None
    """Optional free-form changelog entry."""

    published_by: str = ""
    """Identity of the publisher (node public key hex or human name)."""

    published_at_epoch_ms: int = 0
    """Publication timestamp in milliseconds since Unix epoch."""

    def to_json(self, indent: int | None = 2) -> str:
        """Serialize to a JSON string.

        Args:
            indent: JSON indentation level. None for compact.

        Returns:
            JSON string representation.
        """
        return json.dumps(asdict(self), indent=indent)

    @classmethod
    def from_json(cls, text: str) -> "ContractManifest":
        """Deserialize from a JSON string.

        Args:
            text: JSON string.

        Returns:
            ContractManifest instance.
        """
        data = json.loads(text)
        return cls(**data)

    @classmethod
    def from_file(cls, path: str) -> "ContractManifest":
        """Load a ContractManifest from a JSON file.

        Args:
            path: Path to the manifest JSON file.

        Returns:
            ContractManifest instance.
        """
        with open(path, encoding="utf-8") as f:
            return cls.from_json(f.read())

    def save(self, path: str) -> None:
        """Save the manifest to a JSON file.

        Args:
            path: Destination path for the JSON file.
        """
        with open(path, "w", encoding="utf-8") as f:
            f.write(self.to_json())


# ── FatalContractMismatch ─────────────────────────────────────────────────────


class FatalContractMismatch(Exception):
    """Raised when the live contract hash doesn't match the committed manifest.

    This indicates a breaking change that wasn't recorded. The developer
    must rerun ``aster contract gen`` and commit the updated manifest.
    """

    def __init__(
        self,
        service_name: str,
        version: int,
        expected_id: str,
        actual_id: str,
        manifest_path: str,
    ) -> None:
        self.service_name = service_name
        self.version = version
        self.expected_id = expected_id
        self.actual_id = actual_id
        self.manifest_path = manifest_path

        super().__init__(
            f"Contract identity mismatch for {service_name!r} v{version}:\n"
            f"  Expected: {expected_id}\n"
            f"  Actual:   {actual_id}\n"
            f"  Manifest: {manifest_path}\n"
            f"  → The service interface has changed without updating the manifest.\n"
            f"    Rerun `aster contract gen` and commit the updated manifest."
        )


# ── Verification ──────────────────────────────────────────────────────────────


def verify_manifest_or_fatal(
    live_contract_bytes: bytes,
    manifest_path: str,
) -> ContractManifest:
    """Verify that the live contract bytes match the committed manifest.

    Loads the manifest from *manifest_path*, computes BLAKE3 of
    *live_contract_bytes*, and checks for equality.

    Args:
        live_contract_bytes: Canonical bytes of the live ServiceContract.
        manifest_path: Path to the committed manifest JSON file.

    Returns:
        The loaded ContractManifest on success.

    Raises:
        FatalContractMismatch: If the hashes differ.
        FileNotFoundError: If the manifest file does not exist.
        json.JSONDecodeError: If the manifest file is malformed.
    """
    manifest = ContractManifest.from_file(manifest_path)
    actual_id = blake3.blake3(live_contract_bytes).hexdigest()

    if actual_id != manifest.contract_id:
        raise FatalContractMismatch(
            service_name=manifest.service,
            version=manifest.version,
            expected_id=manifest.contract_id,
            actual_id=actual_id,
            manifest_path=manifest_path,
        )

    return manifest

File: bindings/typescript/packages/aster/src/contract/manifest.ts
/**
 * ContractManifest — persisted contract identity metadata.
 *
 * Spec reference: Aster-ContractIdentity.md S11.4
 *
 * A manifest records a service contract's identity (BLAKE3 hash),
 * method schemas, type hashes, and metadata. Used for:
 * - Startup verification (contract hasn't changed)
 * - Dynamic discovery (shell, MCP, other language clients)
 * - Contract publication (registry + blobs)
 */

import { MAX_MANIFEST_METHODS, MAX_MANIFEST_TYPE_HASHES } from '../limits.js';

/** Method descriptor within a manifest. */
export interface ManifestMethod {
  name: string;
  pattern: string; // "unary" | "server_stream" | "client_stream" | "bidi_stream"
  requestType: string; // hex hash or type name
  responseType: string;
  timeout: number;
  idempotent: boolean;
  /** Request type field descriptors for dynamic invocation. */
  fields: ManifestField[];
  /** Fory wire tags for request/response types. */
  requestWireTag?: string;
  responseWireTag?: string;
  /** Fory wire tags for response fields (for dynamic type synthesis). */
  responseFields?: ManifestField[];
}

/** Field descriptor for dynamic invocation. */
export interface ManifestField {
  name: string;
  type: string; // "str", "int", "float", "bool", "bytes", "list[X]", etc.
  required: boolean;
  default?: unknown;
}

/** Persisted contract manifest. */
export interface ContractManifest {
  service: string;
  version: number;
  contractId: string; // 64-char hex BLAKE3
  canonicalEncoding: string;
  typeCount: number;
  typeHashes: string[]; // hex BLAKE3 per TypeDef
  methodCount: number;
  methods: ManifestMethod[];
  serializationModes: string[];
  scoped: string; // "shared" | "stream"
  deprecated: boolean;
  semver?: string;
  vcsRevision?: string;
  vcsTag?: string;
  vcsUrl?: string;
  changelog?: string;
}

/** Error raised when a live contract doesn't match the manifest. */
export class FatalContractMismatch extends Error {
  constructor(
    public readonly serviceName: string,
    public readonly serviceVersion: number,
    public readonly expectedId: string,
    public readonly actualId: string,
  ) {
    super(
      `Contract identity mismatch for '${serviceName}' v${serviceVersion}:\n` +
      `  expected: ${expectedId}\n` +
      `  actual:   ${actualId}\n` +
      `This usually means the service definition changed without updating the manifest.`,
    );
    this.name = 'FatalContractMismatch';
  }
}

/** Verify that a contract ID matches the manifest. Throws on mismatch. */
export function verifyManifestOrFatal(
  manifest: ContractManifest,
  actualContractId: string,
): void {
  if (manifest.contractId !== actualContractId) {
    throw new FatalContractMismatch(
      manifest.service,
      manifest.version,
      manifest.contractId,
      actualContractId,
    );
  }
}

/** Serialize a manifest to JSON. */
export function manifestToJson(manifest: ContractManifest): string {
  return JSON.stringify({
    service: manifest.service,
    version: manifest.version,
    contract_id: manifest.contractId,
    canonical_encoding: manifest.canonicalEncoding,
    type_count: manifest.typeCount,
    type_hashes: manifest.typeHashes,
    method_count: manifest.methodCount,
    methods: manifest.methods.map(m => ({
      name: m.name,
      pattern: m.pattern,
      request_type: m.requestType,
      response_type: m.responseType,
      timeout: m.timeout,
      idempotent: m.idempotent,
      fields: m.fields,
      request_wire_tag: m.requestWireTag,
      response_wire_tag: m.responseWireTag,
      response_fields: m.responseFields,
    })),
    serialization_modes: manifest.serializationModes,
    scoped: manifest.scoped,
    deprecated: manifest.deprecated,
    semver: manifest.semver,
  }, null, 2);
}

/**
 * Extract method descriptors from a manifest for introspection.
 * Returns a simplified record of method name → pattern.
 */
export function extractMethodDescriptors(manifest: ContractManifest): Record<string, string> {
  const result: Record<string, string> = {};
  for (const method of manifest.methods) {
    result[method.name] = method.pattern ?? 'unary';
  }
  return result;
}

/**
 * Save a manifest to a JSON file.
 */
export function saveManifest(manifest: ContractManifest, filePath: string): void {
  const { writeFileSync } = require('node:fs');
  writeFileSync(filePath, manifestToJson(manifest), 'utf-8');
}

/** Parse a manifest from JSON. */
export function manifestFromJson(json: string): ContractManifest {
  const data = JSON.parse(json);

  // Validate limits
  const methods = data.methods ?? [];
  if (methods.length > MAX_MANIFEST_METHODS) {
    throw new Error(`manifest has ${methods.length} methods, max is ${MAX_MANIFEST_METHODS}`);
  }
  const typeHashes = data.type_hashes ?? [];
  if (typeHashes.length > MAX_MANIFEST_TYPE_HASHES) {
    throw new Error(`manifest has ${typeHashes.length} type hashes, max is ${MAX_MANIFEST_TYPE_HASHES}`);
  }

  // Coerce version to number
  const version = typeof data.version === 'string' ? parseInt(data.version, 10) : data.version;

  return {
    service: data.service,
    version,
    contractId: data.contract_id,
    canonicalEncoding: data.canonical_encoding ?? 'fory-xlang/0.15',
    typeCount: data.type_count ?? 0,
    typeHashes,
    methodCount: data.method_count ?? methods.length,
    methods: methods.map((m: any) => ({
      name: m.name,
      pattern: m.pattern,
      requestType: m.request_type ?? '',
      responseType: m.response_type ?? '',
      timeout: m.timeout ?? 0,
      idempotent: m.idempotent ?? false,
      fields: m.fields ?? [],
      requestWireTag: m.request_wire_tag,
      responseWireTag: m.response_wire_tag,
      responseFields: m.response_fields,
    })),
    serializationModes: data.serialization_modes ?? [],
    scoped: data.scoped ?? 'shared',
    deprecated: data.deprecated ?? false,
    semver: data.semver,
    vcsRevision: data.vcs_revision,
    vcsTag: data.vcs_tag,
    vcsUrl: data.vcs_url,
    changelog: data.changelog,
  };
}
